polygon/bridge/bridge.go

```diff
@@ -33,47 +33,59 @@ import (
 
 	libcommon "github.com/erigontech/erigon-lib/common"
 	"github.com/erigontech/erigon/core/types"
-	"github.com/erigontech/erigon/polygon/bor/borcfg"
 	"github.com/erigontech/erigon/polygon/heimdall"
 )
 
 type eventFetcher interface {
 	FetchStateSyncEvents(ctx context.Context, fromId uint64, to time.Time, limit int) ([]*heimdall.EventRecordWithTime, error)
 }
 
+type DBConfig struct {
+	DataDir   string
+	RoTxLimit int64
+}
+
 type Config struct {
-	DataDir      string
-	Logger       log.Logger
-	BorConfig    *borcfg.BorConfig
-	EventFetcher eventFetcher
-	RoTxLimit    int64
+	Logger                log.Logger
+	StateReceiverContract string
+	EventFetcher          eventFetcher
+
+	IsSprintStartFn           func(uint64) bool
+	CalculateSprintLengthFn   func(uint64) uint64
+	IsIndoreFn                func(uint64) bool
+	CalculateStateSyncDelayFn func(uint64) uint64
+	OverrideStateSyncRecords  map[string]int
 }
 
-func Assemble(config Config) *Bridge {
-	bridgeDB := polygoncommon.NewDatabase(config.DataDir, kv.PolygonBridgeDB, databaseTablesCfg, config.Logger, false /* accede */, config.RoTxLimit)
+func Assemble(config Config, dbConfig DBConfig) *Bridge {
+	bridgeDB := polygoncommon.NewDatabase(dbConfig.DataDir, kv.PolygonBridgeDB, databaseTablesCfg, config.Logger, false /* accede */, dbConfig.RoTxLimit)
 	bridgeStore := NewStore(bridgeDB)
-	reader := NewReader(bridgeStore, config.Logger, config.BorConfig.StateReceiverContract)
-	return NewBridge(bridgeStore, config.Logger, config.BorConfig, config.EventFetcher, reader)
+	reader := NewReader(bridgeStore, config.Logger, config.StateReceiverContract)
+
+	return NewBridge(bridgeStore, config, reader)
 }
 
-func NewBridge(store Store, logger log.Logger, borConfig *borcfg.BorConfig, eventFetcher eventFetcher, reader *Reader) *Bridge {
+func NewBridge(store Store, config Config, reader *Reader) *Bridge {
 	return &Bridge{
 		store:                        store,
-		logger:                       logger,
-		borConfig:                    borConfig,
-		eventFetcher:                 eventFetcher,
-		stateReceiverContractAddress: libcommon.HexToAddress(borConfig.StateReceiverContract),
+		logger:                       config.Logger,
+		eventFetcher:                 config.EventFetcher,
+		stateReceiverContractAddress: libcommon.HexToAddress(config.StateReceiverContract),
 		reader:                       reader,
 		transientErrors:              []error{context.DeadlineExceeded, heimdall.ErrBadGateway},
 		fetchedEventsSignal:          make(chan struct{}),
 		processedBlocksSignal:        make(chan struct{}),
+		isSprintStart:                config.IsSprintStartFn,
```

Comment:
The config can be embedded into the bridge instead of destructuring the config fields.

Reply:
I guess this would go against the whole point of your PR, but what is the problem with :

```type Bridge struct {
     ...
     ...
     *borcfg.BorConfig
     
    ```

Reply:
@antonis19 this is a bad take, why would you want to expose all of the `BorConfig` methods via the `Bridge`? Embedding is useful in other situations, what you are suggesting is a misuse of embedding. Uber's style guide also warns against doing this - https://github.com/uber-go/guide/blob/master/style.md#avoid-embedding-types-in-public-structs

Reply:
My main point was about Bridge using `BorConfig` inside of Bridge instead of defining extra fields. Whether we use Go struct embedding or not is another discussion.  In fact, you were the one that suggested the Go struct embedding in your comment above, which is what I took some inspiration from.

Reply:
I suggested it for `Config` and `DBConfig` - that is a valid use case.
Not for `Bridge` and `BorConfig`.

Reply:
A valid use case according to what? The uber style guide you reference completely discourages the use of embedded types in public structs. (I actually disagree with Uber's style guide on this).

Reply:
@antonis19 a valid use case is when you want to compose attributes and/or functions. 

For the `Config` case we want to compose attributes that are repeating.

For the `Bridge` component case we don't want to compose all of the 20 functions of the `*borcfg.BorConfig` object because we do not want to expose any of these 20 functions as public functions of the `Bridge`.

This is actually pretty simple and straightforward. If you can't see it after this explanation I'm not sure how else I can help you.

Reply:
I don't disagree with you. If you are arguing between using struct embedding or not for BorConfig I agree with you that is better not to use it in this case for the reasons you outlined in your last comment. It was just out of convenience of notation that I wrote it like that, and beside the point that I was making.

Also I think it is completely fine to do the struct embedding in your example.

But you were started arguing the first point using Uber's style guide, which I don't think is a valid supporting argument since that would also argue against your last comment (which I find perfectly valid).